src/twisted/conch/ssh/transport.py

```diff
@@ -637,83 +659,186 @@ def sendPacket(self, messageType, payload):
+            payloadSize += 4
+
         bs = self.currentEncryptions.encBlockSize
-        # 4 for the packet length and 1 for the padding length
-        totalSize = 5 + len(payload)
-        lenPad = bs - (totalSize % bs)
+        lenPad = bs - (payloadSize % bs)
         if lenPad < 4:
             lenPad = lenPad + bs
-        packet = (
-            struct.pack("!LB", totalSize + lenPad - 4, lenPad)
-            + payload
-            + randbytes.secureRandom(lenPad)
+
+        # Create the unencrypted packet.
+        payloadPacket = (
+            struct.pack("!B", lenPad) + payload + randbytes.secureRandom(lenPad)
+        )
+
+        if isETM:
+            # For Encrypt-then-MAC the total size is not encrypted.
+            encryptedPayload = self.currentEncryptions.encrypt(payloadPacket)
+            totalSizePacket = struct.pack("!L", len(encryptedPayload))
+            macTarget = totalSizePacket + encryptedPayload
+            macPacket = self.currentEncryptions.makeMAC(
+                self.outgoingPacketSequence, macTarget
+            )
+            return macTarget + macPacket
+
+        # For non ETM, we subtract the 4 bytes added to compute the
+        # padding size.
+        # But we also encrypt the package length.
+        totalSizePacket = struct.pack("!L", payloadSize + lenPad - 4)
+        encryptedPacket = self.currentEncryptions.encrypt(
+            totalSizePacket + payloadPacket
+        )
+        macTarget = totalSizePacket + payloadPacket
+
+        macPacket = self.currentEncryptions.makeMAC(
+            self.outgoingPacketSequence, macTarget
         )
-        encPacket = self.currentEncryptions.encrypt(
-            packet
-        ) + self.currentEncryptions.makeMAC(self.outgoingPacketSequence, packet)
-        self.transport.write(encPacket)
-        self.outgoingPacketSequence += 1
 
-    def getPacket(self):
+        return encryptedPacket + macPacket
+
+    def getPacket(self) -> bytes | None:
         """
         Try to return a decrypted, authenticated, and decompressed packet
         out of the buffer.  If there is not enough data, return None.
 
-        @rtype: L{str} or L{None}
-        @return: The decoded packet, if any.
+        @return: The decoded packet, if any. C{None} is returned if no packet was received yet.
         """
-        bs = self.currentEncryptions.decBlockSize
-        ms = self.currentEncryptions.verifyDigestSize
-        if len(self.buf) < bs:
+        if len(self.buf) < self.currentEncryptions.decBlockSize:
             # Not enough data for a block
             return
+
+        try:
+            if self.currentEncryptions.isInMACETM():
+                payload = self._getPacketETM()
+            else:
+                payload = self._getPacketMTE()
+
+        except _InvalidPacket as error:
+            self.sendDisconnect(
+                DISCONNECT_PROTOCOL_ERROR,
+                error.args[0].encode("ascii"),
+            )
+            return
```

Comment:
If we are always discarding the traceback from this new exception, it might be better expressed as an addition to the union; less stack-walking that way, more explicit that it's a possibility.  i.e. `payload: bytes | None | Literal[Invalid.packet]` and then `if packet is Invalid.packet:` ?

Reply:
We don't keep the traceback, but we extract the error message.

I think that what is missing here, is ` self._log.failure(error.args[0])`.

This should make it similar to the other error exception handling.

---------

I prefer to use the exception here as I would like to have the error message, and I feel that raising an exceptoin is the "standard" way of signaling an invalid packet while also having an error details.

-------

From my experience, the whole log calls from the SSH transport are not well designed.

For my production code, I ignore them and I have various custom hooks to add logs.

Improving the logs would be a different tasks for conch.ssh